### PR TITLE
[misc] use multi-stage build in Dockerfile

### PR DESCRIPTION
**Pull request by IJtLJZ8Rm4Yr**

### CODE REVIEW

**Review by fenrir-z — APPROVED** [review]

File: ymir/Dockerfile.backend [hunk 1]
```diff
@@ -1,3 +1,11 @@
+FROM golang:1.16-buster AS build
```
Comment (phoenix-xhuang) [inline ymir/Dockerfile.backend:1]: 使用1.16版本是因为ubuntu 20.04自带最高版本只有1.16，用multistage就可以用当前最新稳定版本1.18.5

File: ymir/Dockerfile.backend [hunk 2]
```diff
@@ -1,3 +1,11 @@
...
+
+COPY ./backend/src /app
```
Comment (phoenix-xhuang) [inline ymir/Dockerfile.backend:3]: 需要/backend/src目录吗？

File: ymir/Dockerfile.backend [hunk 3]
```diff
@@ -24,18 +32,13 @@ RUN mkdir -p /app_logs/
 COPY ./backend/requirements.txt ./
```
Comment (phoenix-xhuang) [inline ymir/Dockerfile.backend:32]: 看上去前面的golang-1.16-go可以不需要了
“go build command compiles the packages, along with their dependencies. By default go build will generate an executable for the current [platform and architecture](https://www.digitalocean.com/community/tutorials/building-go-applications-for-different-operating-systems-and-architectures#possible-platforms-for-goos-and-goarch). For example, if built on a linux/386 system, the executable will be compatible with any other linux/386 system, even if Go is not installed.”

File: ymir/Dockerfile.backend [hunk 4]
```diff
@@ -1,3 +1,11 @@
+FROM golang:1.16-buster AS build
+
```
Comment (phoenix-xhuang) [inline ymir/Dockerfile.backend:2]: 需要在copy完整代码前，加上以下代码，先下载依赖包。否则每次代码改动就要重新下载一次依赖
```
FROM golang:1.18.5-buster AS build

WORKDIR /app
COPY ./backend/src/ymir_viewer/go.mod .
COPY ./backend/src/ymir_viewer/go.sum .
RUN go mod download -x

COPY ./backend/src/ymir_viewer/ /ymir_viewer
WORKDIR /ymir_viewer
RUN go build -mod=readonly -v -o viewer_server
```

**Comment by IJtLJZ8Rm4Yr** [comment]
use https://github.com/IndustryEssentials/ymir/pull/1021 instead

**Pull request closed without merging**